### PR TITLE
Remove deep validation and instead use error hierarchy to improve error messages

### LINKED ISSUE #2175: Chart.encode() returns confusing error message when using an invalid channel and selections
Consider:

```python
selection = alt.selection_single()
(alt.Chart(data=None)
    .mark_circle()
    .encode(color=alt.value('red'), invalidChannel=None))
```

Predictably, this fails with:

```
SchemaValidationError: Invalid specification
        altair.vegalite.v4.schema.core.FacetedEncoding, validating 'additionalProperties'
        Additional properties are not allowed ('invalidChannel' was unexpected)
```

So far so good. But now watch what happens if `x` is set to a selection:

```python
selection = alt.selection_single()
(alt.Chart(data=None)
    .mark_circle()
    .add_selection(selection)
    .encode(
        color=alt.condition(selection, alt.value('red'), alt.value('green')),
        invalidChannel=None))
```

This time, the error message is way more confusing and doesn't point to the actual cause of the problem:

```
SchemaValidationError: Invalid specification
        altair.vegalite.v4.schema.channels.ColorValue, validating 'additionalProperties'
        Additional properties are not allowed ('selection' was unexpected)
```

The misleading error message made me waste a lot of time chasing down dead ends simply because I mistyped `tooltip` as `tooltips` :(

### LINKED ISSUE #2609: Improve the error message when trying to use `.properties` on concatenated and facet charts
I think it would be great to provide a better error message when `properties` are used on a concatenated or layered chart. Currently this results in `Data is a required property` which is not really informative since the error is that the chart dimensions need to be specified on the individual charts. I think this is especially confusing since it works for layered charts.

So this works:

```python
import altair as alt
from vega_datasets import data

source = data.cars.url

points = alt.Chart(source).mark_point().encode(
    x='Horsepower:Q',
    y='Miles_per_Gallon:Q',
)
(points + points).properties(width=400)
```

But this raises the error:

```python
(points | points).properties(width=400)
```

A better error message for this was already suggested in https://github.com/altair-viz/altair/issues/2495#issuecomment-926839178. An alternative solution would be to allow `properties` on concatenated/faceted charts to maintain syntactical parity with layered charts and then silently set it on each chart under the hood, but there might be caveats with this so the improved error message seems like the easiest way forward.

Is `properties` ever relevant for concatenated/faceted charts or could we just add a message such as ".properties are not allowed for concatenated or faceted charts, please set it on each individual chart instead." similar to the messages for layering of concatenated/faceted charts?

### LINKED ISSUE #2744: Incorrect error message when an unknown param value is used
I just noted that in a spec that contains a condition, using an unknown parameter value anywhere in the spec will always return an error pointing to the condition even when it has nothing to do with actual error. For example, the following spec:

```py
import altair as alt
from vega_datasets import data

source = data.barley()

alt.Chart(source).mark_bar().encode(
    x='variety',
    y=alt.Y('sum(yield)', stack='asdf'),
    opacity=alt.condition('datum.yield > 0', alt.value(1), alt.value(0.2))
)
```

returns:

```
SchemaValidationError: Invalid specification

        altair.vegalite.v4.schema.channels.OpacityValue, validating 'additionalProperties'

        Additional properties are not allowed ('test' was unexpected)
```

If we change to `opacity=alt.value(0.6)` we get the expected output:

```
SchemaValidationError: Invalid specification

        altair.vegalite.v4.schema.channels.Y->0, validating 'enum'

        '' is not one of ['zero', 'center', 'normalize']
```

Note that there is nothing wrong the opacity condition here and removing the incorrect `stack` value leads to zero errors. Looking closer at the error raising sequence, it seems like the stack error is always raised first, but if there is a conditional in the spec there will also be a second error raised. Not sure how raising one error leads to the other or why the sequence doesn't stop at the first raised error.

It seems like `{'test': 'datum.yield > 0', 'value': 1}` is evaluating against a schema that only contains the `condition` key, so it should have been `{'condition': {'test': 'datum.yield > 0', 'value': 1}}` instead and then `{'test': 'datum.yield > 0', 'value': 1}` in the next validation.

### PR DESCRIPTION
**Pull request by binste**
# Background
Many exception messages that a user sees in Altair are based on the validation of the chart specification against the vega-lite schema. This validation with the package `jsonschema` produces error messages on multiple level of the schema, e.g. if an error happens in a subschema, it produces errors for the subschema as well as the schemas above. For the user, the most informative errors are the one directly pinpointing the root cause, i.e. the errors on the lowest level where the wrong specification occurs.

To generate and raise these "deep" errors, #555 introduced "deep validation". This meant that a chart is validated once using `jsonschema`, if a validation error is raised, the chart is validated again but this time the validation is performed iteratively on all individual subschemas in a recursive fashion. This was a great improvement and I'm not sure if there was a better way at that time. However, these subschemas by themselves are not always valid, see #2609 for an example, which in turn leads to new confusing error messages which have nothing to do with the original issue.

# Proposed solution
This PR removes the deep validation again. Instead, it only validates the chart specifications once and then uses the exception hierarchy which is returned by `jsonschema` to find the root cause. See the "Walkthrough" section below for more details. With this approach, we can still raise the exception which is at the root of the issue, but there is no need to validate individual schemas and therefore issues such as #2609 are solved:
```python
import altair as alt
from vega_datasets import data

source = data.cars.url

points = alt.Chart(source).mark_point().encode(
    x='Horsepower:Q',
    y='Miles_per_Gallon:Q',
)
(points | points).properties(width=400)
```
raises correctly:
```python
SchemaValidationError: Invalid specification

        altair.vegalite.v5.api.HConcatChart, validating 'additionalProperties'

        Additional properties are not allowed ('width' was unexpected)
```

Furthermore, we can combine exceptions on the same "hierarchy level" in the schema to provide even more messages:

```python
alt.Chart(data.barley()).mark_bar().encode(
    x=alt.X('variety'),
    y=alt.Y('sum(yield)', stack='asdf'),
)
```
```python
SchemaValidationError: Invalid specification

        altair.vegalite.v5.api.Chart->0, validating 'enum'

        'asdf' is not one of ['zero', 'center', 'normalize']
        'asdf' is not of type 'null'
        'asdf' is not of type 'boolean'
```
Previously, the error only showed the line `'asdf' is not one of ['zero', 'center', 'normalize']`, although `stack` could also be a boolean or null.

I already added these examples and a few more to the tests to make sure that informative exceptions are raised. Let me know if you have more that I can test against!

Fixes #2175, fixes #2609, fixes #2744, related PRs which are probably no longer needed: #2841, #2835
I tested the code also against jsonschema 3.0 which is the minimum version required by Altair.

# Walkthrough
To help with the review process, I try to outline below how we get from an invalid chart specification to the final error message which is raised. I can't put in links to all the places in the code I refer to as they are not modified in the PR. Let me know if you need more details.

Let's start with the example from above:

* The validation of a chart object is still triggered in the `TopLevelMixin.to_dict` method which then ends up calling `super(...).to_dict` [here](https://github.com/altair-viz/altair/pull/2842/files#diff-bbe4e187b18d242a366c820d023afd89041759cb96e4ec66c3f34559a72c2f9dR559) which propagates to `SchemaBase.to_dict`
* `SchemaBase.to_dict` calls `self.validate` which uses the `validate_jsonschema` function to validate the specification. This raises a `jsonschema.ValidationError`. This error is catched in `SchemaBase.to_dict` and turned into a custom `SchemaValidationError` which has better error messages and then raised again
* Without this PR. This error would read `{'aggregate': 'sum', 'field': 'yield', 'stack': 'asdf', 'type': 'quantitative'} is not valid under any of the given schemas` which is not very informative. This is why the deep validation was originally introduced.
* This error would be [catched in `TopLevelMixin.to_dict`](https://github.com/altair-viz/altair/pull/2842/files#diff-bbe4e187b18d242a366c820d023afd89041759cb96e4ec66c3f34559a72c2f9dL560) and then the deep validation is run iterating through all objects in the specification and validating them individually. This raises the final error `'asdf' is not one of ['zero', 'center', 'normalize']`to the user which is much better.

With the new approach implemented in this PR, we use the error hierarchy which is in the `context` attribute of the original `jsonschema.ValidationError` which was raised. *"If the error was caused by errors in subschemas, the list of errors from the subschemas will be available on this property."*, see [the jsonschema docs](https://python-jsonschema.readthedocs.io/en/stable/errors/) for all the details. These errors from the subschemas have again a `context` attribute which allows to find the root error. I added a new function [_get_most_relevant_errors](https://github.com/altair-viz/altair/pull/2842/files#diff-25906c2570b127f0bb9cff4550bcc153006b2fdb36aea55ae804ca9a74eb4404R79) which tries to find the most relevant errors based on this hierarchy. For the example it looks like this:

* The originally raised error is: `ValidationError: "{'aggregate': 'sum', 'field': 'yield', 'stack': 'asdf', 'type': 'quantitative'} is not valid under any of the given schemas` and from now on referenced with `top_error`
* `top_error.context` gives
  ```python
  [<ValidationError: "'asdf' is not valid under any of the given schemas">,
   <ValidationError: "Additional properties are not allowed ('aggregate', 'field' were unexpected)">,
   <ValidationError: "'asdf' is not valid under any of the given schemas">,
   <ValidationError: "Additional properties are not allowed ('aggregate', 'field', 'stack', 'type' were unexpected)">,
   <ValidationError: "'value' is a required property">]
  ```
* The function always takes the first error in the context and sees if this is already the lowest level, i.e. if there is another `context` or if its `None`. In this case, there is more. `top_error.context[0].context` gives:
  ```python
  [<ValidationError: "'asdf' is not one of ['zero', 'center', 'normalize']">,
   <ValidationError: "'asdf' is not of type 'null'">,
   <ValidationError: "'asdf' is not of type 'boolean'">]
  ```
* Calling `.context` again on the first validation error in this list shows that it is None and therefore we reached the bottom of the hierarchy. 
* The old deep validation approach would have raised this error alone. We can now use all 3 errors to craft a more informative error message. `_get_most_relevant_errors` will return all 3 errors. `validate_jsonschema` will then raise the first of them (`ValidationError: "'asdf' is not one of ['zero', 'center', 'normalize']"`) and add the other two as `_additional_errors` attribute to the first. For errors of type `additionalProperties` only the first error on the lowest level is enough as the others do not contain more information, they are instead more verbose and somewhat confusing messages.
* This error that is now raised is still converted, as in the old approach, to a `SchemaValidationError` so it should not break any users code if they have `except` statements. We [use the 2 additional errors](https://github.com/altair-viz/altair/pull/2842/files#diff-25906c2570b127f0bb9cff4550bcc153006b2fdb36aea55ae804ca9a74eb4404R175) to craft the final error message 

```python
SchemaValidationError: Invalid specification

        altair.vegalite.v5.api.Chart->0, validating 'enum'

        'asdf' is not one of ['zero', 'center', 'normalize']
        'asdf' is not of type 'null'
        'asdf' is not of type 'boolean'
```

### CODE REVIEW

**Comment by mattijn** [comment]
Great work @binste! @joelostblom is probably more capable in reviewing this than myself. 
One question though, maybe not specific on this PR, but more in general on error messages based on schema such as:

```json
'asdf' is not of type 'null'
```

In Python Altair, `None` is used instead of `'null'`. There may be other instances of this. Is it necessary to translate these instances if they appear in an error message?

**Comment by ChristopherDavisUCI** [comment]
@mattijn Is this dictionary possibly relevant to your comment?

https://github.com/altair-viz/altair/blob/a8172b7134ed262224e5dfb4e0c99d4bdc55bcb0/tools/schemapi/codegen.py#L226-L234

(In another version, in `tools/schemapi/utils.py`, it was `mapping` instead of `dict`.  Let me know if you think it's a mistake that I am using `dict` in this context.)

**Comment by binste** [comment]
I think translating those types would be very helpful to users! Thanks @ChristopherDavisUCI, definitely looks relevant. There is also https://github.com/altair-viz/altair/blob/a8172b7134ed262224e5dfb4e0c99d4bdc55bcb0/tools/schemapi/utils.py#L183 which is only slightly different. I'm happy to tackle this as part of a larger effort around types as mentioned in https://github.com/altair-viz/altair/pull/2846#issuecomment-1404066238 but if anyone gets a chance earlier to replace the types in the error messages then please, don't hold back! :)

**Comment by joelostblom** [comment]
In short, this is looking fantastic! Thank you so much for digging into the depths of this @binste and sorry for the delay in reviewing. I am going to get to this in the next couple of days.

On brief comment now, do you think it is confusing to see three error messages like in your example above? I don't know what would be a feasible solution technically here, but I am thinking that instead of

```
        'asdf' is not one of ['zero', 'center', 'normalize']
        'asdf' is not of type 'null'
        'asdf' is not of type 'boolean'
```

the message would be clearer as:

```
        'asdf' needs to be one of ['zero', 'center', 'normalize'] or of type 'null' or boolean'
```

This is a minor comment though

**Comment by binste** [comment]
No worries and thank you for the kind words! :)

Indeed, it would be a nice enhancement if the error messages can be merged into a one-liner. For that we probably need to check how consistent these strings are and if that changes across versions for `jsonschema`. Or maybe we can construct it ourselves from some other underlying attributes of the exceptions. I made a note for myself to create a new issue with your suggestion once this PR is merged so we can look at it separately in detail.

File: altair/utils/schemapi.py
Comment (joelostblom) [inline altair/utils/schemapi.py:362]: I find it confusing that there are two parameters names `validate`, one in the private method `_to_dict` which we manually set to `False` everywhere in this PR, and one in `to_dict` which is actually controlled by the user. Could we remove the `validate` param of the  `_todict` private function and just pass `False` on line 129 where `to_dict` is called inside `_todict`? It seems like this will never be `True` with the changes in this PR.

**Review by joelostblom — CHANGES_REQUESTED** [review]
As I said above, this PR is fantastic and makes both the code logic and the error message much more straightforward! One additional benefit of this PR is that it makes the traceback **much** shorter. Before, it was common to have around 120 lines or so of traceback; now that is down to 20, which makes it much easier to find the error message without having to scroll so much.

Also a big thanks for the details comments in the code and explanations in the PR which made it much easier to review. I have a few minor comments and then we're good to merge.

File: altair/utils/schemapi.py
Comment (binste) [reply altair/utils/schemapi.py:429]: You're right, thanks!

File: altair/utils/schemapi.py
Comment (binste) [reply altair/utils/schemapi.py:83]: Yes that's true. I thought it's nice if it works also with empty sequences in case this function is ever going to be used anywhere else in the codebase. Easier to build in the behaviour now then later. But I can also see that it's currently redundant so happy to remove it if you'd still prefer that.

File: tests/utils/tests/test_schemapi.py
Comment (binste) [reply tests/utils/tests/test_schemapi.py:435]: No specific reason for the lambdas and good point about the clearer overview. Adjusted it accordingly.

File: tests/utils/tests/test_schemapi.py
Comment (binste) [reply tests/utils/tests/test_schemapi.py:432]: Without space is the style pytest uses in [their documentation](https://docs.pytest.org/en/6.2.x/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions). I think this is the first occurence in Altair tests with multiple arguments. Both works so I don't mind to adopt a style with space :)

File: altair/utils/schemapi.py
Comment (binste) [reply altair/utils/schemapi.py:175]: I think I'll first want to do some further testing under which conditions multiple errors can appear and how the errors are phrased and if that message could be better constructed (in a separate PR) based on the information `jsonschema` returns.

File: altair/utils/schemapi.py
Comment (binste) [reply altair/utils/schemapi.py:362]: Good point, done.

**Comment by binste** [comment]
All comments except two are implemented. For those I'll wait for your feedback.

File: tests/utils/tests/test_schemapi.py
Comment (joelostblom) [reply tests/utils/tests/test_schemapi.py:466]: There are actually a few places that use this style in Altair already and they seem to use a space. For example:

https://github.com/altair-viz/altair/blob/8a8642b2e7eeee3b914850a8f7aacd53335302d9/tools/schemapi/tests/test_utils.py#L36-L40

and 

https://github.com/altair-viz/altair/blob/97ff1ebc3a485be1c00a34e5d19801eca5287e6f/tests/vegalite/v5/tests/test_api.py#L179-L183

Personally I don't really mind which style we use, but it would be nice to be consistent (although honestly I think this is very minor too so let me know what you prefer and we can merge this PR afterwards).

**Review by joelostblom — APPROVED** [review]

File: tests/utils/tests/test_schemapi.py
Comment (binste) [reply tests/utils/tests/test_schemapi.py]: Ah sorry, not sure how I missed those. I like the consistency as well so I committed your code suggestion. Good to go from my side, thanks for the review @joelostblom!

**Comment by binste** [comment]
Thanks for the review! I also added a line to the changelog. Not sure if this is the preferred way moving forward to keep it up-to-date? Could be nice to include a line for every PR which is relevant for the changelog but let me know if I should remove it again.

**Comment by joelostblom** [comment]
Yup I agree, I think each notable PR should come with a changelog entry so that we can easier keep track. Will merge when checks go green.

**Comment by mattijn** [comment]
Whoop whoop. Congrats @binste for this one!

**Comment by binste** [comment]
Happy that this approach worked out, thanks! :)

**Pull request merged**
